File: server/src/services/import_persist.py
"""Shared two-phase persistence skeleton for the image importers.

Both ``danbooru_import`` and ``gallery_dl_import`` land new posts + their tags
the same way: normalise each source record into a :class:`NormalizedPostRow`,
then run two short, independently-retried transactions on a *worker-thread*
SQLite connection — tags first, posts + ``post_has_tag`` second. The only real
difference between the two callers is where the row fields come from (a
``DanbooruPost`` vs a ``GalleryDLItem``), so that mapping stays in each importer
and everything below is shared.

Persist always happens AFTER the files are on disk (see the importer docstrings):
a post row must never exist before its file, or the sync reconciler's
``remove_deleted_files`` races the throttled download and deletes the
just-committed post mid-flight. This module is downstream of that ordering — it
only ever receives rows whose bytes already landed.
"""

from __future__ import annotations

import contextlib
import sqlite3
from dataclasses import dataclass, field
from typing import TYPE_CHECKING, Any

from shared import logger

if TYPE_CHECKING:
    from collections.abc import Callable
    from datetime import datetime


@dataclass
class NormalizedPostRow:
    """One image ready to persist, in the ``posts`` table's own terms.

    The fields are exactly the ``INSERT INTO posts(...)`` column list both
    importers already used (``file_path``, ``file_name``, ``extension``,
    ``source``, ``rating``, ``published_at``) plus the post's ``{tag_name:
    group_id}`` map. Callers do all source-specific work — id stringification,
    source resolution, rating coercion, tag/group flattening — before building
    this, so the persistence core stays format-agnostic.
    """

    file_path: str
    file_name: str
    extension: str
    source: str
    rating: int
    # danbooru hands sqlite3 a datetime (bound via the default adapter, as it
    # always did); gallery-dl passes an already-formatted string.
    published_at: str | datetime | None
    tags: dict[str, int] = field(default_factory=dict)


def persist_posts_with_tags(db: Any, rows: list[NormalizedPostRow]) -> list[int]:
    """Persist ``rows`` + their tags in two transactions; return the post ids.

    Called on a pool worker thread — ``db.cursor()`` therefore returns a cursor
    on *this worker thread's* SQLite connection, not the event-loop thread's.
    That isolation matters: transactions in sqlite3 are connection-scoped, so if
    all concurrent requests shared one connection, one worker's ``_safe_rollback``
    could (and did) rip a sibling worker's in-flight BEGIN out from under it —
    the ``cannot commit - no transaction is active`` failure mode.

    Split rationale: when concurrent import requests all insert overlapping tags,
    the commit-time uniqueness check on ``tags(name)`` aborts one of them. Running
    tag inserts in their own short transaction keeps that retry surface tiny and
    prevents replay of the (much larger) posts + post_has_tag work each time tags
    happen to conflict.

    Each transaction uses ON CONFLICT for in-snapshot duplicates and a bounded
    retry loop for commit-time conflicts that only show up against rows committed
    by other transactions after our snapshot was taken.
    """
    if not rows:
        return []
    cur = db.cursor()
    try:
        # Phase A: globally-deduped tag upsert in its own short transaction.
        all_tags: dict[str, int] = {}
        for row in rows:
            for name, gid in row.tags.items():
                all_tags.setdefault(name, gid)
        if all_tags:
            _run_with_retry(cur, "tags", lambda: _insert_tags_tx(cur, all_tags))

        # Phase B: posts + post_has_tag in their own transaction. The tags they
        # reference are now committed by phase A, so concurrent writers can't
        # make this transaction wait on them. ``result`` is reassigned per
        # attempt so only the successful run's ids survive a mid-flight retry.
        result: dict[str, list[int]] = {"ids": []}

        def _insert_posts() -> None:
            result["ids"] = _insert_posts_tx(cur, rows)

        _run_with_retry(cur, "posts", _insert_posts)
        return result["ids"]
    finally:
        cur.close()


def _run_with_retry(
    cur: sqlite3.Cursor,
    label: str,
    fn: Callable[[], None],
    *,
    max_attempts: int = 5,
) -> None:
    """Retry on SQLite ``database is locked`` while another writer holds it.

    With WAL mode and a single backend process the writer lock is short-lived,
    but the startup backfill task can collide with import requests; retry a few
    times before giving up.
    """
    for attempt in range(1, max_attempts + 1):
        try:
            fn()
        except sqlite3.OperationalError as exc:
            _safe_rollback(cur)
            msg = str(exc).lower()
            if "locked" not in msg and "busy" not in msg:
                raise
            if attempt == max_attempts:
                raise
            logger.warning(
                f"Import {label} write contention (attempt {attempt}/{max_attempts}): {exc}; retrying",
            )
        except Exception:
            _safe_rollback(cur)
            raise
        else:
            return


def _insert_tags_tx(cur: sqlite3.Cursor, all_tags: dict[str, int]) -> None:
    cur.execute("BEGIN")
    cur.executemany(
        "INSERT INTO tags(name, group_id) VALUES (?, ?) ON CONFLICT(name) DO NOTHING",
        list(all_tags.items()),
    )
    cur.execute("COMMIT")


def _insert_posts_tx(cur: sqlite3.Cursor, rows: list[NormalizedPostRow]) -> list[int]:
    cur.execute("BEGIN")
    post_tag_pairs: list[tuple[int, dict[str, int]]] = []
    for row in rows:
        cur.execute(
            """
            INSERT INTO posts(file_path, file_name, extension, source, rating, published_at)
            VALUES (?, ?, ?, ?, ?, ?)
            ON CONFLICT (file_path, file_name, extension)
            DO UPDATE SET source = excluded.source,
                          published_at = excluded.published_at,
                          updated_at = CURRENT_TIMESTAMP
            RETURNING id
            """,
            [row.file_path, row.file_name, row.extension, row.source, row.rating, row.published_at],
        )
        # RETURNING fires for both INSERT (new post) and the DO UPDATE branch
        # (a pre-existing tag-less row being backfilled), so we get the id
        # either way and the post_has_tag upsert below attaches the tags.
        fetched = cur.fetchone()
        if fetched:
            post_tag_pairs.append((int(fetched[0]), row.tags))

    # (post_id, tag_name) is unique within this batch — each post_id appears
    # once and per-post names were deduped via dict when the row was built.
    post_tag_rows = [(post_id, name) for post_id, tag_map in post_tag_pairs for name in tag_map]
    if post_tag_rows:
        cur.executemany(
            "INSERT INTO post_has_tag(post_id, tag_name, is_auto) VALUES (?, ?, 0) ON CONFLICT DO NOTHING",
            post_tag_rows,
        )
    cur.execute("COMMIT")
    return [post_id for post_id, _ in post_tag_pairs]


def _safe_rollback(cur: sqlite3.Cursor) -> None:
    """ROLLBACK that swallows the 'no transaction is active' case."""
    with contextlib.suppress(sqlite3.OperationalError):
        cur.execute("ROLLBACK")

File: packages/db/src/repositories/import-persist.ts
/**
 * 导入器共用的两段式落库 —— 对应 Python 侧 `services/import_persist.py`。
 *
 * Danbooru 和 gallery-dl 落新 post 的方式是同一套：把每条源记录规整成
 * `NormalizedRow`，然后跑两个短事务 —— **先 tags，再 posts + post_has_tag**。
 *
 * 分成两个事务不是洁癖：并发导入插入重叠的 tag 时，`tags(name)` 的唯一性检查会在
 * 提交时把其中一个打回。让 tag 插入自己占一个短事务，重试面就只有那一小块，而不用
 * 把（大得多的）posts + post_has_tag 那一堆跟着重放一遍。
 *
 * ⚠️ 落库永远发生在**文件已经落盘之后**（见导入器的注释）：post 行绝不能早于它的
 * 文件存在，否则 sync 的对账会把它当成"文件已删除"，在下载还没跑完时就删掉它。
 * 这个模块处在那个顺序的下游 —— 它收到的行，字节都已经在磁盘上了。
 */
import type BetterSqlite3 from 'better-sqlite3'

export interface NormalizedRow {
  filePath: string
  fileName: string
  extension: string
  source: string
  rating: number
  publishedAt: string | null
  /** `{tag_name: group_id}`。 */
  tags: Record<string, number>
}

/**
 * 落 `rows` 和它们的 tag，返回 post id。
 *
 * `ON CONFLICT DO UPDATE` 那一支同样会触发 `RETURNING`，所以一条**已经存在但没有
 * tag** 的裸行（磁盘对账建出来的那种）也能拿到 id 并被补上 tag —— 这正是重跑一次
 * 导入能把它们补全的原因。
 */
export function persistPostsWithTags(
  sqlite: BetterSqlite3.Database,
  rows: NormalizedRow[],
): number[] {
  if (!rows.length)
    return []

  // 阶段 A：整批去重后的 tag upsert，自己一个短事务。
  const allTags = new Map<string, number>()
  for (const row of rows) {
    for (const [name, gid] of Object.entries(row.tags)) {
      if (!allTags.has(name))
        allTags.set(name, gid)
    }
  }
  if (allTags.size) {
    const insTag = sqlite.prepare(
      'INSERT INTO tags(name, group_id) VALUES (?, ?) ON CONFLICT(name) DO NOTHING',
    )
    sqlite.transaction(() => {
      for (const [name, gid] of allTags) insTag.run(name, gid)
    })()
  }

  // 阶段 B：posts + post_has_tag。它们引用的 tag 已经被 A 提交了。
  const insPost = sqlite.prepare<unknown[], { id: number }>(
    'INSERT INTO posts(file_path, file_name, extension, source, rating, published_at) '
    + 'VALUES (?, ?, ?, ?, ?, ?) '
    + 'ON CONFLICT (file_path, file_name, extension) '
    + 'DO UPDATE SET source = excluded.source, published_at = excluded.published_at, '
    + 'updated_at = CURRENT_TIMESTAMP '
    + 'RETURNING id',
  )
  const insLink = sqlite.prepare(
    'INSERT INTO post_has_tag(post_id, tag_name, is_auto) VALUES (?, ?, 0) ON CONFLICT DO NOTHING',
  )

  const ids: number[] = []
  sqlite.transaction(() => {
    for (const row of rows) {
      const got = insPost.get(
        row.filePath,
        row.fileName,
        row.extension,
        row.source,
        row.rating,
        row.publishedAt,
      )
      if (!got)
        continue
      ids.push(got.id)
      for (const name of Object.keys(row.tags)) insLink.run(got.id, name)
    }
  })()
  return ids
}

/**
 * 某个目录下**已经带 tag 导入过**的 Danbooru post id。
 *
 * id 从 `posts.file_name` 里来，因为对这个导入器来说那一列**就是** Danbooru 的
 * post id。去重过滤和翻页停止条件都拿原始 post id 和这个集合比，所以这个等价关系
 * 是承重的。
 *
 * ⚠️ 判据是"已经有一个手工（`is_auto = 0`）标签"，**不是**"post 行存在"。一个文件
 * 可能先以无标签的形式进库 —— 磁盘对账会给任何已在盘上的文件插一条裸行，而一次
 * DB 重置或快照回滚也会留下没有 `post_has_tag` 的文件。按"行存在"去重会让那些裸行
 * 永远被跳过（它们的 file_name 就在那儿），Danbooru 的标签永远写不进去。
 */
export function listImportedDanbooruIds(
  sqlite: BetterSqlite3.Database,
  filePathStr: string,
): string[] {
  return sqlite
    .prepare<[string], { file_name: string }>(
      'SELECT p.file_name FROM posts p '
      + 'JOIN post_has_tag pht ON pht.post_id = p.id AND pht.is_auto = 0 '
      + 'WHERE p.file_path = ?',
    )
    .all(filePathStr)
    .map(r => r.file_name)
}
